### ISSUE TITLE
TPI fixes

### ISSUE THREAD

**Issue by jdebacker** [jdebacker/ProblemSet9#43]
Here are a few things to look to fix in `TPI.py` (not necessarily an exhaustive list, but some that jump out at me):

- [ ] The distance calculation for the guessed and implied paths should look at deviations for the BQ path and the r path
- [ ] There should be updating of the r path just like the BQ path is updated in lines 68-70
- [ ] When updating the paths of r and BQ, just update through period T (though you will need the arrays to be of length T+S-1 to solve the full time path (which includes those born in period T)
- [ ] The defining of initial paths (lines 21-34 -- or at least the parts of these that set the initial BQ and r paths) should be outside the while loop
- [ ] The time path solution should return aggregate labor over the path, since that is one other variables you'll need to plot

**Comment by miltonash** [comment]
Thanks for the feedback! I’ll continue working with `TPI.py` tonight.

**Comment by xinxinzhang123** [comment]
- To update r, I  try following cods based on Miltion's codes. Are these codes right?

       # use market clearing
        L_path = np.ones(T) * agg.get_L(n, omega_SS)
        # find implied r
        new_rpath = firm.get_r(L_path, new_Kpath, r_params)
        # update r
        rpath_init[:T] = xi * new_rpath[:T] + (1 - xi) * rpath_init[:T]

- You mentioned that the distance calculation for the guessed and implied paths should look at deviations for the BQ path and the r path. Is the following distance code right?
 dist = ((BQpath_init[:T] - new_BQpath[:T]) ** 2).sum()+ ((rpath_init[:T] - new_rpath[:T]) ** 2).sum()

**Comment by jdebacker** [comment]
@xinxinzhang123 Yes, that code snippet looks good to me.  Your distance calculation should also work, just be careful how you set the tolerance so that it's appropriate for the distance norm you use.   But you can always adjust the tolerance parameter after you get things running.
